include/nbl/builtin/hlsl/workgroup/arithmetic.hlsl

```diff
@@ -0,0 +1,127 @@
+// Copyright (C) 2023 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_WORKGROUP_ARITHMETIC_INCLUDED_
+#define _NBL_BUILTIN_HLSL_WORKGROUP_ARITHMETIC_INCLUDED_
+
+#include "nbl/builtin/hlsl/cpp_compat.hlsl"
+#include "nbl/builtin/hlsl/workgroup/ballot.hlsl"
+#include "nbl/builtin/hlsl/workgroup/broadcast.hlsl"
+#include "nbl/builtin/hlsl/workgroup/shared_scan.hlsl"
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace workgroup
+{
+
+#define REDUCE Reduce<T, subgroup::inclusive_scan<T, Binop>, SharedAccessor>
+#define SCAN(isExclusive) Scan<T, Binop, subgroup::inclusive_scan<T, Binop>, SharedAccessor, isExclusive>
+template<typename T, class Binop, class SharedAccessor>
+T reduction(T value, NBL_REF_ARG(SharedAccessor) accessor)
+{
+    REDUCE reduce = REDUCE::create(_NBL_HLSL_WORKGROUP_SIZE_);
+    reduce(value, accessor);
+    accessor.main.workgroupExecutionAndMemoryBarrier();
+    T retVal = Broadcast<uint, SharedAccessor>(reduce.lastLevelScan, accessor, reduce.lastInvocationInLevel);
+    return retVal;
+}
+
+template<typename T, class Binop, class SharedAccessor>
+T inclusive_scan(T value, NBL_REF_ARG(SharedAccessor) accessor)
+{
+    SCAN(false) incl_scan = SCAN(false)::create(_NBL_HLSL_WORKGROUP_SIZE_, Binop::identity());
+    T retVal = incl_scan(value, accessor);
+    return retVal;
+}
+
+template<typename T, class Binop, class SharedAccessor>
+T exclusive_scan(T value, NBL_REF_ARG(SharedAccessor) accessor)
+{
+    SCAN(true) excl_scan = SCAN(true)::create(_NBL_HLSL_WORKGROUP_SIZE_, Binop::identity());
+    T retVal = excl_scan(value, accessor);
+    return retVal;
+}
+
+#undef REDUCE
+#undef SCAN
+
+#define REDUCE Reduce<uint, subgroup::inclusive_scan<uint, binops::add<uint> >, SharedAccessor>
+#define SCAN Scan<uint, binops::add<uint>, subgroup::inclusive_scan<uint, binops::add<uint> >, SharedAccessor, true>
+/**
+ * Gives us the sum (reduction) of all ballots for the workgroup.
+ *
+ * Only the first few invocations are used for performing the sum 
+ * since we only have `uballotBitfieldCount` amount of uints that we need 
+ * to add together.
+ * 
+ * We add them all in the shared array index after the last DWORD 
+ * that is used for the ballots. For example, if we have 128 workgroup size,
+ * then the array index in which we accumulate the sum is `4` since 
+ * indexes 0..3 are used for ballots.
+ */ 
+template<class SharedAccessor>
+uint ballotBitCount(NBL_REF_ARG(SharedAccessor) accessor)
+{
+    uint participatingBitfield;
+    if(gl_LocalInvocationIndex < uballotBitfieldCount)
+    {
+        participatingBitfield = accessor.ballot.get(gl_LocalInvocationIndex);
+    }
+    accessor.ballot.workgroupExecutionAndMemoryBarrier();
+    REDUCE wsh = REDUCE::create(uballotBitfieldCount);
```

Comment:
REDUCE and SCAN taking a runtime item count prevents two things:
- static assert when item count > Workgroup Size (cause its dumb to do and code won't work anyway
- the compiler aggressively unrolling and optimizing

I'm not against "smaller" scan sizes, but make sure the size is a compile time constant in a template.